### DISCUSSION TITLE
Update about J2CL compatibility?

### DISCUSSION THREAD

**Discussion by salmonb** [foal/gwt-time#16] [Q&A]
hi,

Thanks a lot for your library, which I use in my [WebFX](https://github.com/webfx-project/webfx) framework, a JavaFX transpiler powered by GWT, and I'm currently investigating the possibility to move to J2CL (actually to support both GWT & J2CL).

I noticed in your README that you had a plan to make your library J2CL compatible. I'm interested indeed. And apparently there is no such library for J2CL so far.

Can you please give an update of your progress?

Thank you!

**Comment by foal** [comment]
Well, really it maybe J2CL compatible, but nobody tests it. 

The main issue is super-source processing. Current implementation has several "tricks" that allows to compile the library by usual Java as well as GWT compiler. I am not sure that same tricks are suitable for J2CL.

You can find them in the POM files it is:
 - package relocation - from org.jresearch.threetenbp.gwt.emu -> java.time. Reason: Java 11 does not allow to use java.* for own classes.
 - stub generation - some files have different implementation for Java and GWT (check @Wrap annotation)

**Reply by salmonb** [reply]
Thanks for your reply,

I can be your tester! 😄

I will probably do some tests by the end of the week, otherwise beginning of next week.

I will let you know how it goes

**Comment by salmonb** [comment]
So I started the tests, and couldn't make it work with j2cl so far... 🤷

The thing is that there is no more `.gwt.xml` modules with j2cl, it just compiles all what you put in the dependencies (actually through the [J2CL Maven plugin](https://github.com/Vertispan/j2clmavenplugin) I'm using here). So if I include your module, it will try to compile everything inside, not just the super sources, and that's not what we want (it fails with errors anyway).

So I think you need to provide a `-j2cl` version of it, with only the super sources.

This is what I tried to do in this [repo1](https://github.com/webfx-project/gwt-time-j2cl) (just for testing purposes - you can run `mvn install` locally).
And then I made this minimal j2cl test [repo2](https://github.com/webfx-project/gwt-time-j2cl-test), but I probably missed something  in repo1, because the j2cl build (that you can trigger with `mvn clean package`) fails as follow:
````

...
[INFO] Starting dev.webfx:gwt-time-j2cl:0.1.0-SNAPSHOT/stripped_bytecode_headers
[INFO] dev.webfx:gwt-time-j2cl:0.1.0-SNAPSHOT/stripped_bytecode_headers: /Users/bruno/IdeaProjects/gwt-time-j2cl-test/target/gwt3BuildCache/0.22.0/dev.webfx-gwt-time-j2cl-0.1.0-SNAPSHOT/602bc2426247880097428e15ec54de78-stripped_sources/results/super/java/time/Duration.java:59: error: symbol not found org.gwtproject.regexp.shared.RegExp
import org.gwtproject.regexp.shared.RegExp;
       ^
/Users/bruno/IdeaProjects/gwt-time-j2cl-test/target/gwt3BuildCache/0.22.0/dev.webfx-gwt-time-j2cl-0.1.0-SNAPSHOT/602bc2426247880097428e15ec54de78-stripped_sources/results/super/java/time/Duration.java:114: error: could not resolve RegExp
    private final static RegExp PATTERN =
                         ^
...
````
 
Maybe you can use repo1 & repo2 to investigate further, or just let me know what I need to correct.

Also let me know if you prefer I open an issue rather than using this discussion.

Thank you!

**Reply by foal** [reply]
Right now, I am on vacation. 

I think we can remove the JUnit tests from J2CL package. It has to solve 90% issue with J2CL compatible.

Thank you for the help!

**Reply by salmonb** [reply]
No pb.

Since your library needs a bit of work to make it J2CL compatible, I finally opened an issue #17, as I think it's better to continue there.

Enjoy the rest of your vacation.
